File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/PopupMenuPreference.java
package org.totschnig.myexpenses.preference;

import android.content.Context;
import android.util.AttributeSet;
import android.view.Menu;
import android.view.View;

import androidx.appcompat.widget.PopupMenu;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

public class PopupMenuPreference extends Preference {
  private View anchorView;
  public PopupMenuPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    super(context, attrs, defStyleAttr, defStyleRes);
  }

  public PopupMenuPreference(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  public PopupMenuPreference(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public PopupMenuPreference(Context context) {
    super(context);
  }

  @Override
  public void onBindViewHolder(PreferenceViewHolder holder) {
    super.onBindViewHolder(holder);
    anchorView = holder.itemView;
  }

  public void showPopupMenu(PopupMenu.OnMenuItemClickListener listener, String... items) {
    PopupMenu popup = new PopupMenu(getContext(), anchorView);
    Menu popupMenu = popup.getMenu();
    popup.setOnMenuItemClickListener(listener);
    for (int i = 0; i < items.length; i++) {
      popupMenu.add(Menu.NONE, i, Menu.NONE, items[i]);
    }
    popup.show();
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/PopupMenuPreference.kt
package org.totschnig.myexpenses.preference

import android.content.Context
import android.util.AttributeSet
import android.view.Menu
import android.view.View
import androidx.appcompat.widget.PopupMenu
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder

class PopupMenuPreference(context: Context, attrs: AttributeSet) : Preference(context, attrs) {
    private lateinit var anchorView: View

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        anchorView = holder.itemView
    }

    fun showPopupMenu(vararg items: String, listener: PopupMenu.OnMenuItemClickListener) {
        showPopupMenu(
            populateMenu = {
                items.forEachIndexed { index, s ->
                    it.add(Menu.NONE, index, Menu.NONE, s)
                }
            },
            listener
        )
    }

    fun showPopupMenu(populateMenu: (Menu) -> Unit, listener: PopupMenu.OnMenuItemClickListener) {
        PopupMenu(context, anchorView).apply {
            setOnMenuItemClickListener(listener)
            populateMenu(menu)
            show()
        }
    }
}